rfd/00XY-session-locking.md

```diff
@@ -0,0 +1,269 @@
+    // A matching node is also unregistered and prevented from registering.
+    string Node;
+    
+    // MFADevice specifies the ID of an MFA device recorded in a user certificate.
+    string MFADevice;
+}
+```
+
+In most cases `SessionLockTarget` should list just a single attribute.
+(It would be preferable to define it as a `oneof`, however setting a custom
+`jsontag` for `oneof` is [not supported by `gogoproto`](https://github.com/gogo/protobuf/issues/623).)
+
+Note that `SessionLock` is not a singleton resource: when there are multiple
+`SessionLock`s stored (and in force), it suffices for a session to be matched
+by any one of the locks to be terminated/disabled.
+
+#### Relation to `User.Status`
+
+There already exists a field of `User` resource that is used to
+capture a user lock in connection with failed Web UI login attempts.
+
+This `Status` field and its `LoginStatus` definition are superseded by
+`SessionLock`.  All of its use cases should be converted to `SessionLock`.
+
+The `SessionLock` approach allows to specify locks for entities that are only
+yet to exist or exist merely transiently (such as SSO user objects).  It could
+also help with alleviating the load associated with caching `User` resources.
+
+#### `tctl` support
+
+`SessionLock` resources can be managed using `tctl [get|create|rm]`.  In this
+way, it is possible to update or remove a session lock after it
+has been created.
+
+There will be a special `tctl sessions lock` helper provided, to facilitate
+supplying time information when creating new `SessionLock`s, see Scenarios below.
+
+### Propagation within a Teleport cluster
+
+For the purposes of lock detection and evaluation, the `SessionLock` resource will be propagated using the cache system to all Teleport roles/services.
+
+An `auth.AccessPoint` wrapper will be introduced that makes the getters return a `trace.ConnectionProblem` for stale data. The wrapper will be parametrized by a duration that defines the maximum failure period after which the data are considered stale.
+
+`SessionLockTarget` will be convertible into a filter for `types.WatchKind`, allowing
+the creation of watchers that match specific contexts. The `GetSessionLocks` method
+should also accept `SessionLockTarget`s as filters.
+
+### Fallback mode
+
+If the "strict" cache returns an error indicating stale data, there is a decision to be made about whether to rely on the last known locks. There will be two levels on which to determine this decision.
+
+1. If a lockable transaction involves a user and the user's certificate has an extension named `teleport-session-lock-fallback`, the transaction is locked out based on the extension's value which comes from the user's RBAC role:
+
+```
+spec:
+   options: 
+       # In strict mode, if the session locks are not up to date the session will be terminated.
+       # In best_effort mode, the most recent view of the locks will remain to be used.
+       session_lock_fallback: [strict|best_effort]
```

Comment:
Nit on naming: `session_lock: strict | best_effort`. Fallback part is not necessary - seems like an internal detail.